File: app/services/sumup_service.py
import uuid
from typing import Any

from flask import current_app
from sumup import APIError, Sumup
from sumup.checkouts import CreateCheckoutBody


class SumUpService:
    def __init__(self, api_key: str | None = None) -> None:
        self.api_key = api_key or current_app.config.get("SUMUP_API_KEY")
        self.client = Sumup(api_key=self.api_key)

    def create_checkout(
        self,
        amount: int,
        currency: str = "EUR",
        description: str = "",
        checkout_reference: str = "",
        merchant_code: str | None = None,
    ) -> dict[str, Any] | None:
        try:
            # Get merchant code from config if not provided
            if not merchant_code:
                merchant_code = current_app.config.get("SUMUP_MERCHANT_CODE")

            # Merchant code is required by SumUp SDK
            if not merchant_code:
                current_app.logger.error("SUMUP_MERCHANT_CODE not configured")
                raise ValueError("SUMUP_MERCHANT_CODE must be set in environment variables")

            # Generate checkout reference if not provided
            if not checkout_reference:
                checkout_reference = str(uuid.uuid4())

            # Convert cents to euros for SumUp API
            amount_euros = amount / 100.0

            # Create checkout body
            checkout_body = CreateCheckoutBody(
                amount=float(amount_euros),
                currency=currency,
                checkout_reference=checkout_reference,
                merchant_code=merchant_code,
                description=description,
            )

            # Create checkout using SDK
            response = self.client.checkouts.create(body=checkout_body)

            # Log the response for debugging
            current_app.logger.info(f"SumUp checkout created with ID: {response.id if hasattr(response, 'id') else 'no id'}")

            # The SDK returns a Checkout object
            if response and hasattr(response, "id"):
                checkout_id = response.id

                return {
                    "id": checkout_id,
                    "checkout_reference": (response.checkout_reference if hasattr(response, "checkout_reference") else checkout_reference),
                    "status": (response.status if hasattr(response, "status") else "PENDING"),
                }

            current_app.logger.error("SumUp checkout response missing id")
            return None

        except APIError as e:
            current_app.logger.error(f"SumUp API error creating checkout: {str(e)}")
            return None
        except ValueError as e:
            current_app.logger.error(f"Configuration error: {str(e)}")
            return None
        except Exception as e:
            current_app.logger.error(f"Error creating SumUp checkout: {str(e)}")
            return None

    def get_checkout(self, checkout_id: str) -> Any | None:
        try:
            checkout = self.client.checkouts.get(checkout_id=checkout_id)
            return checkout
        except APIError as e:
            current_app.logger.error(f"SumUp API error getting checkout: {str(e)}")
            return None
        except Exception as e:
            current_app.logger.error(f"Error getting SumUp checkout: {str(e)}")
            return None

    def verify_payment(self, checkout_id: str) -> bool:
        try:
            checkout = self.get_checkout(checkout_id)

            if not checkout:
                return False

            # Check if checkout status is PAID
            status = checkout.status if hasattr(checkout, "status") else None
            return status == "PAID"

        except Exception as e:
            current_app.logger.error(f"Error verifying SumUp payment: {str(e)}")
            return False

File: app/Services/SumUpService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Str;

class SumUpService
{
    private string $apiKey;

    private string $merchantCode;

    private string $baseUrl = 'https://api.sumup.com/v0.1';

    public function __construct()
    {
        $this->apiKey = (string) config('services.sumup.api_key', '');
        $this->merchantCode = (string) config('services.sumup.merchant_code', '');
    }

    /**
     * Create a checkout for payment.
     *
     * @return array<string, mixed>|null
     */
    public function createCheckout(
        int $amountCents,
        string $currency = 'EUR',
        string $description = '',
        ?string $checkoutReference = null,
    ): ?array {
        if (empty($this->apiKey) || empty($this->merchantCode)) {
            Log::error('SumUp API key or merchant code not configured');

            return null;
        }

        $checkoutReference = $checkoutReference ?? (string) Str::uuid();
        $amountEuros = $amountCents / 100.0;

        try {
            $response = Http::withToken($this->apiKey)
                ->post("{$this->baseUrl}/checkouts", [
                    'checkout_reference' => $checkoutReference,
                    'amount' => $amountEuros,
                    'currency' => $currency,
                    'merchant_code' => $this->merchantCode,
                    'description' => $description,
                ]);

            if ($response->successful()) {
                $data = $response->json();
                Log::info('SumUp checkout created', ['id' => $data['id'] ?? 'no id']);

                return [
                    'id' => $data['id'] ?? null,
                    'checkout_reference' => $data['checkout_reference'] ?? $checkoutReference,
                    'status' => $data['status'] ?? 'PENDING',
                ];
            }

            Log::error('SumUp API error creating checkout', [
                'status' => $response->status(),
                'body' => $response->body(),
            ]);

            return null;
        } catch (\Exception $e) {
            Log::error('Error creating SumUp checkout', ['error' => $e->getMessage()]);

            return null;
        }
    }

    /**
     * Get a checkout by ID.
     *
     * @return array<string, mixed>|null
     */
    public function getCheckout(string $checkoutId): ?array
    {
        try {
            $response = Http::withToken($this->apiKey)
                ->get("{$this->baseUrl}/checkouts/{$checkoutId}");

            if ($response->successful()) {
                return $response->json();
            }

            Log::error('SumUp API error getting checkout', [
                'checkout_id' => $checkoutId,
                'status' => $response->status(),
            ]);

            return null;
        } catch (\Exception $e) {
            Log::error('Error getting SumUp checkout', ['error' => $e->getMessage()]);

            return null;
        }
    }

    /**
     * Verify if a payment has been completed.
     */
    public function verifyPayment(string $checkoutId): bool
    {
        $checkout = $this->getCheckout($checkoutId);

        if (! $checkout) {
            return false;
        }

        return ($checkout['status'] ?? '') === 'PAID';
    }

    /**
     * Process a checkout payment with card details.
     *
     * @return array<string, mixed>
     */
    public function processCheckoutPayment(
        string $checkoutId,
        string $cardNumber,
        string $cardName,
        string $expiryMonth,
        string $expiryYear,
        string $cvv,
    ): array {
        try {
            // Format expiry year to 4 digits
            if (strlen($expiryYear) === 2) {
                $expiryYear = '20' . $expiryYear;
            }

            // Format expiry month to 2 digits
            if (strlen($expiryMonth) === 1) {
                $expiryMonth = '0' . $expiryMonth;
            }

            $response = Http::withToken($this->apiKey)
                ->put("{$this->baseUrl}/checkouts/{$checkoutId}", [
                    'payment_type' => 'card',
                    'card' => [
                        'number' => str_replace(' ', '', $cardNumber),
                        'name' => $cardName,
                        'expiry_month' => $expiryMonth,
                        'expiry_year' => $expiryYear,
                        'cvv' => $cvv,
                    ],
                ]);

            $data = $response->json();
            $status = $data['status'] ?? 'FAILED';

            Log::info('SumUp payment processed', [
                'checkout_id' => $checkoutId,
                'status' => $status,
            ]);

            return [
                'success' => $status === 'PAID',
                'status' => $status,
                'checkout_id' => $checkoutId,
                'transaction_id' => $data['transaction_code'] ?? $data['transaction_id'] ?? null,
                'transaction_code' => $data['transaction_code'] ?? null,
            ];
        } catch (\Exception $e) {
            Log::error('Error processing SumUp payment', ['error' => $e->getMessage()]);

            return [
                'success' => false,
                'status' => 'FAILED',
                'error' => $e->getMessage(),
            ];
        }
    }
}
